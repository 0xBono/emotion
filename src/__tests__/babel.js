/* eslint-disable jsx-quotes,no-useless-escape,no-template-curly-in-string */
/* eslint-env jest */
import plugin from '../babel'

const babel = require('babel-core')

describe('emotion/babel', () => {
  describe('babel styled component', () => {
    test('basic', () => {
      const basic = 'emotion.h1`font-size: ${fontSize}px;`'
      const { code } = babel.transform(basic, {
        plugins: [plugin, 'glam/babel']
      })
      expect(code).toMatchSnapshot()
    })

    test('function call', () => {
      const basic = 'emotion(MyComponent)`font-size: ${fontSize}px;`'
      const { code } = babel.transform(basic, {
        plugins: [plugin, 'glam/babel']
      })
      expect(code).toMatchSnapshot()
    })

    test('attr', () => {
      const basic = `styled('input')\`
       margin: attr(margin);
       color: #ffffff;
       height: \$\{props => props.height * props.scale\};
       width: attr(width);
       color: blue;
       display: \$\{flex\};
      \``
      const { code } = babel.transform(basic, {
        plugins: [plugin, 'glam/babel']
      })
      expect(code).toMatchSnapshot()
    })

    test('attr with value type', () => {
      const basic = `styled('input')\`
        margin: attr(margin px);
      \``
      const { code } = babel.transform(basic, {
        plugins: [plugin, 'glam/babel']
      })
      expect(code).toMatchSnapshot()
    })

    test('attr with default value', () => {
      const basic = `styled('input')\`
        margin: attr(margin, 16);
      \``
      const { code } = babel.transform(basic, {
        plugins: [plugin, 'glam/babel']
      })
      expect(code).toMatchSnapshot()
    })

    test('attr with value type and default value', () => {
      const basic = `styled('input')\`
        margin: attr(margin px, 16);
      \``
      const { code } = babel.transform(basic, {
        plugins: [plugin, 'glam/babel']
      })
      expect(code).toMatchSnapshot()
    })

    test('match works on multiple', () => {
      const basic = `styled('input')\`
        margin: attr(margin px, 16);
        color: blue;
        padding: attr(padding em, 16);
      \``
      const { code } = babel.transform(basic, {
        plugins: [plugin, 'glam/babel']
      })
      expect(code).toMatchSnapshot()
    })

    test('attr kitchen sink', () => {
      const basic = `styled('input')\`
        margin: attr(margin px, 16);
        padding: attr(padding em, 16);
        font-size: attr(fontSize ch, 8);
        width: attr(width %, 95);
        height: attr(height vw, 90);
        display: attr(display, flex);
      \``
      const { code } = babel.transform(basic, {
        plugins: [plugin, 'glam/babel']
      })
      expect(code).toMatchSnapshot()
    })

    test('basic inline', () => {
      const basic = 'emotion.h1`font-size: ${fontSize}px;`'
      const { code } = babel.transform(basic, {
        plugins: [
          plugin,
          [
            'glam/babel',
            {
              sync: true,
              inline: true
            }
          ]
        ]
      })
      expect(code).toMatchSnapshot()
    })

    test('inline with attr', () => {
      const basic = `styled('input')\`
        margin: attr(margin px, 16);
        padding: attr(padding em, 16);
        font-size: attr(fontSize ch, 8);
        width: attr(width %, 95);
        height: attr(height vw, 90);
        display: attr(display, flex);
      \``
      const { code } = babel.transform(basic, {
        plugins: [plugin, ['glam/babel', { sync: true, inline: true }]]
      })
      expect(code).toMatchSnapshot()
    })
  })

  describe('babel css prop', () => {
    test('basic', () => {
      const basic = '(<div className="a" css={`color: brown;`}></div>)'
      const { code } = babel.transform(basic, {
        plugins: [plugin, 'glam/babel']
      })
      expect(code).toMatchSnapshot()
    })

    test('basic inline', () => {
      const basic = '(<div className="a" css={`color: brown;`}></div>)'
      const {code} = babel.transform(basic, {
        plugins: [plugin, ['glam/babel', { inline: true }]]
      })
      expect(code).toMatchSnapshot()
    })

    test('dynamic inline', () => {
      const basic = '(<div className="a" css={`color: ${color};`}></div>)'
      const {code} = babel.transform(basic, {
        plugins: [plugin, ['glam/babel', {inline: true}]]
      })
      expect(code).toMatchSnapshot()
    })

    test('no css attr', () => {
      const basic = '(<div></div>)'
      const { code } = babel.transform(basic, {
        plugins: [plugin, 'glam/babel']
      })
      expect(code).toMatchSnapshot()
    })

    test('css empty', () => {
      const basic = '(<div css=""></div>)'
      const { code } = babel.transform(basic, {
        plugins: [plugin, 'glam/babel']
      })
      expect(code).toMatchSnapshot()
    })

    test('wrong value type', () => {
      const basic = '(<div css={5}></div>)'
      expect(() =>
        babel.transform(basic, { plugins: [plugin, 'glam/babel'] })
      ).toThrow()
    })

    test('StringLiteral css prop value', () => {
      const basic = `<div css="color: brown;"></div>`
      const { code } = babel.transform(basic, {
        plugins: [plugin, 'glam/babel']
      })
      expect(code).toMatchSnapshot()
    })

    test('noClassName', () => {
      const basic = '(<div css={`color: brown;`}></div>)'
      const { code } = babel.transform(basic, {
        plugins: [plugin, 'glam/babel']
      })
      expect(code).toMatchSnapshot()
    })

    test('emptyClassName', () => {
      const basic = '(<div className="" css={`color: brown;`}></div>)'
      const { code } = babel.transform(basic, {
        plugins: [plugin, 'glam/babel']
      })
      expect(code).toMatchSnapshot()
    })

    test('className as expression', () => {
      const basic = '(<div className={variable} css={`color: brown;`}></div>)'
      const { code } = babel.transform(basic, {
        plugins: [plugin, 'glam/babel']
      })
      expect(code).toMatchSnapshot()
    })

    test('className as expression string', () => {
      const basic =
        '(<div className={`test__class`} css={`color: brown;`} this={`hello`}></div>)'
      const { code } = babel.transform(basic, {
        plugins: [plugin, 'glam/babel']
      })
      expect(code).toMatchSnapshot()
    })
  })
})
