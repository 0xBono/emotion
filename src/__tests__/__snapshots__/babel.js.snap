// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`emotion/babel babel css prop StringLiteral css prop value 1`] = `"<div className={css(\\"css-jf1v9l\\", [])}></div>;"`;

exports[`emotion/babel babel css prop basic 1`] = `"<div className={\\"a\\" + \\" \\" + css(\\"css-jf1v9l\\", [])}></div>;"`;

exports[`emotion/babel babel css prop basic inline 1`] = `
"<div className={\\"a\\" + \\" \\" + css(\\"css-jf1v9l\\", [], function inlineCss() {
  return [\`.css-jf1v9l { color: brown; }\`];
})}></div>;"
`;

exports[`emotion/babel babel css prop className as expression 1`] = `"<div className={variable + \\" \\" + css(\\"css-jf1v9l\\", [])}></div>;"`;

exports[`emotion/babel babel css prop className as expression string 1`] = `"<div className={\`test__class\` + \\" \\" + css(\\"css-jf1v9l\\", [])} this={\`hello\`}></div>;"`;

exports[`emotion/babel babel css prop css empty 1`] = `"<div className={css(\\"css-0\\", [])}></div>;"`;

exports[`emotion/babel babel css prop dynamic inline 1`] = `
"<div className={\\"a\\" + \\" \\" + css(\\"css-jsqoaa\\", [color], function inlineCss(x0) {
  return [\`.css-jsqoaa { color: \${x0}; }\`];
})}></div>;"
`;

exports[`emotion/babel babel css prop emptyClassName 1`] = `"<div className={\\"\\" + \\" \\" + css(\\"css-jf1v9l\\", [])}></div>;"`;

exports[`emotion/babel babel css prop no css attr 1`] = `"<div></div>;"`;

exports[`emotion/babel babel css prop noClassName 1`] = `"<div className={css(\\"css-jf1v9l\\", [])}></div>;"`;

exports[`emotion/babel babel emotion component attr 1`] = `
"emotion('input', ['css-14vghnb', [function getMargin(props) {
       return props.margin;
}, props => props.height * props.scale, function getWidth(props) {
       return props.width;
}, flex]]);"
`;

exports[`emotion/babel babel emotion component attr kitchen sink 1`] = `
"emotion('input', ['css-1lqib20', [function getMargin(props) {
        return props.margin ? props.margin : ('16' + 'px');
}, function getPadding(props) {
        return props.padding ? props.padding : ('16' + 'em');
}, function getWidth(props) {
        return props.width ? props.width : ('95' + '%');
}, function getHeight(props) {
        return props.height ? props.height : ('90' + 'vw');
}, function getDisplay(props) {
        return props.display ? props.display : ('flex' + '');
}]]);"
`;

exports[`emotion/babel babel emotion component attr with default value 1`] = `
"emotion('input', ['css-1luxokk', [function getMargin(props) {
        return props.margin ? props.margin : ('16' + '');
}]]);"
`;

exports[`emotion/babel babel emotion component attr with value type 1`] = `
"emotion('input', ['css-1luxokk', [function getMargin(props) {
        return props.margin + 'px';
}]]);"
`;

exports[`emotion/babel babel emotion component attr with value type and default value 1`] = `
"emotion('input', ['css-1luxokk', [function getMargin(props) {
        return props.margin ? props.margin : ('16' + 'px');
}]]);"
`;

exports[`emotion/babel babel emotion component basic 1`] = `"emotion(\\"h1\\", [\\"css-8xpzga\\", [fontSize]]);"`;

exports[`emotion/babel babel emotion component basic inline 1`] = `
"emotion(\\"h1\\", [\\"css-8xpzga\\", [fontSize], function inlineCss(x0) {
  return [\`.css-8xpzga { font-size: \${x0}px; }\`];
}]);"
`;

exports[`emotion/babel babel emotion component function call 1`] = `"emotion(MyComponent, [\\"css-8xpzga\\", [fontSize]]);"`;

exports[`emotion/babel babel emotion component inline with attr 1`] = `
"emotion('input', ['css-1lqib20', [function getMargin(props) {
        return props.margin ? props.margin : ('16' + 'px');
}, function getPadding(props) {
        return props.padding ? props.padding : ('16' + 'em');
}, function getWidth(props) {
        return props.width ? props.width : ('95' + '%');
}, function getHeight(props) {
        return props.height ? props.height : ('90' + 'vw');
}, function getDisplay(props) {
        return props.display ? props.display : ('flex' + '');
}], function inlineCss(x0, x1, x2, x3, x4) {
        return [\`.css-1lqib20 { margin: \${x0};
        padding: \${x1};
        font-size: attr(fontSize ch, 8);
        width: \${x2};
        height: \${x3};
        display: \${x4}; }\`];
}]);"
`;

exports[`emotion/babel babel emotion component match works on multiple 1`] = `
"emotion('input', ['css-1j6h68f', [function getMargin(props) {
        return props.margin ? props.margin : ('16' + 'px');
}, function getPadding(props) {
        return props.padding ? props.padding : ('16' + 'em');
}]]);"
`;
